src/core/lib/channel/call_tracer.cc

```diff
@@ -48,4 +57,284 @@ absl::string_view ServerCallTracerFactory::ChannelArgName() {
+  void AddTracer(ClientCallTracer* tracer) { tracers_.push_back(tracer); }
+
+ private:
+  std::vector<ClientCallTracer*> tracers_;
+};
+
+class DelegatingServerCallTracer : public ServerCallTracer {
+ public:
+  explicit DelegatingServerCallTracer(ServerCallTracer* tracer)
+      : tracers_{tracer} {}
+  ~DelegatingServerCallTracer() override {}
+  void RecordSendInitialMetadata(
+      grpc_metadata_batch* send_initial_metadata) override {
+    for (auto* tracer : tracers_) {
+      tracer->RecordSendInitialMetadata(send_initial_metadata);
+    }
+  }
+  void RecordSendTrailingMetadata(
+      grpc_metadata_batch* send_trailing_metadata) override {
+    for (auto* tracer : tracers_) {
+      tracer->RecordSendTrailingMetadata(send_trailing_metadata);
+    }
+  }
+  void RecordSendMessage(const SliceBuffer& send_message) override {
+    for (auto* tracer : tracers_) {
+      tracer->RecordSendMessage(send_message);
+    }
+  }
+  void RecordSendCompressedMessage(
+      const SliceBuffer& send_compressed_message) override {
+    for (auto* tracer : tracers_) {
+      tracer->RecordSendCompressedMessage(send_compressed_message);
+    }
+  }
+  void RecordReceivedInitialMetadata(
+      grpc_metadata_batch* recv_initial_metadata) override {
+    for (auto* tracer : tracers_) {
+      tracer->RecordReceivedInitialMetadata(recv_initial_metadata);
+    }
+  }
+  void RecordReceivedMessage(const SliceBuffer& recv_message) override {
+    for (auto* tracer : tracers_) {
+      tracer->RecordReceivedMessage(recv_message);
+    }
+  }
+  void RecordReceivedDecompressedMessage(
+      const SliceBuffer& recv_decompressed_message) override {
+    for (auto* tracer : tracers_) {
+      tracer->RecordReceivedDecompressedMessage(recv_decompressed_message);
+    }
+  }
+  void RecordCancel(grpc_error_handle cancel_error) override {
+    for (auto* tracer : tracers_) {
+      tracer->RecordCancel(cancel_error);
+    }
+  }
+  void RecordReceivedTrailingMetadata(
+      grpc_metadata_batch* recv_trailing_metadata) override {
+    for (auto* tracer : tracers_) {
+      tracer->RecordReceivedTrailingMetadata(recv_trailing_metadata);
+    }
+  }
+  void RecordEnd(const grpc_call_final_info* final_info) override {
+    for (auto* tracer : tracers_) {
+      tracer->RecordEnd(final_info);
+    }
+  }
+  void RecordAnnotation(absl::string_view annotation) override {
+    for (auto* tracer : tracers_) {
+      tracer->RecordAnnotation(annotation);
+    }
+  }
+  void RecordAnnotation(const Annotation& annotation) override {
+    for (auto* tracer : tracers_) {
+      tracer->RecordAnnotation(annotation);
+    }
+  }
+  std::string TraceId() override { return tracers_[0]->TraceId(); }
+  std::string SpanId() override { return tracers_[0]->SpanId(); }
+  bool IsSampled() override { return tracers_[0]->IsSampled(); }
+  bool IsDelegatingTracer() override { return true; }
+
+  void AddTracer(ServerCallTracer* tracer) { tracers_.push_back(tracer); }
```

Comment:
Needs something like 'AddTracer cannot be called after the tracer has been passed to the transport' or some such thing

Reply:
Added comments in call_tracer.h. For the client-side, it is straightforward. For the server-side, we really only want the call tracers to be inserted when all the server tracer factories are invoked. I'm still thinking where to put the list of factories (one solution is to use ServerBuilderPlugins that will put in the factories in a list in a new channel arg, but it makes me wonder whether the ServerBuilderPlugin infrastructure is adding any value here). Anyway I would want that in a separate PR so that eyes can specifically look at that.

Reply:
Let's chat soon... I'm beginning to be in favor of having channel tracers be a channel/server level configuration that's independent of filters/interceptors.

Reply:
alright, i'll schedule something for next week